x-pack/plugins/stack_connectors/server/connector_types/inference/inference.ts

```diff
@@ -0,0 +1,233 @@
+
+import { PassThrough, Stream } from 'stream';
+import { IncomingMessage } from 'http';
+
+import { AxiosError } from 'axios';
+import {
+  InferenceInferenceRequest,
+  InferenceInferenceResponse,
+  InferenceTaskType,
+} from '@elastic/elasticsearch/lib/api/types';
+import {
+  ChatCompleteParamsSchema,
+  RerankParamsSchema,
+  SparseEmbeddingParamsSchema,
+  TextEmbeddingParamsSchema,
+} from '../../../common/inference/schema';
+import {
+  Config,
+  Secrets,
+  ChatCompleteParams,
+  ChatCompleteResponse,
+  StreamingResponse,
+  RerankParams,
+  RerankResponse,
+  SparseEmbeddingParams,
+  SparseEmbeddingResponse,
+  TextEmbeddingParams,
+  TextEmbeddingResponse,
+} from '../../../common/inference/types';
+import { SUB_ACTION } from '../../../common/inference/constants';
+
+export class InferenceConnector extends SubActionConnector<Config, Secrets> {
+  // Not using Axios
+  protected getResponseErrorMessage(error: AxiosError): string {
+    throw new Error('Method not implemented.');
+  }
+
+  private inferenceId;
+  private taskType;
+
+  constructor(params: ServiceParams<Config, Secrets>) {
+    super(params);
+
+    this.provider = this.config.provider;
+    this.taskType = this.config.taskType;
+    this.inferenceId = this.config.inferenceId;
+    this.logger = this.logger;
+    this.connectorID = this.connector.id;
+    this.connectorTokenClient = params.services.connectorTokenClient;
+
+    this.registerSubActions();
+  }
+
+  private registerSubActions() {
+    this.registerSubAction({
+      name: SUB_ACTION.COMPLETION,
+      method: 'performApiCompletion',
+      schema: ChatCompleteParamsSchema,
+    });
+
+    this.registerSubAction({
+      name: SUB_ACTION.RERANK,
+      method: 'performApiRerank',
+      schema: RerankParamsSchema,
+    });
+
+    this.registerSubAction({
+      name: SUB_ACTION.SPARSE_EMBEDDING,
+      method: 'performApiSparseEmbedding',
+      schema: SparseEmbeddingParamsSchema,
+    });
+
+    this.registerSubAction({
+      name: SUB_ACTION.TEXT_EMBEDDING,
+      method: 'performApiTextEmbedding',
+      schema: TextEmbeddingParamsSchema,
+    });
+
+    this.registerSubAction({
+      name: SUB_ACTION.COMPLETION_STREAM,
+      method: 'performApiCompletionStream',
+      schema: ChatCompleteParamsSchema,
+    });
+  }
+
+  /**
+   * responsible for making a esClient inference method to perform chat completetion task endpoint and returning the service response data
+   * @param input the text on which you want to perform the inference task.
+   * @signal abort signal
+   */
+  public async performApiCompletion({
+    input,
+    signal,
+  }: ChatCompleteParams & { signal?: AbortSignal }): Promise<ChatCompleteResponse> {
+    const response = await this.performInferenceApi(
+      { inference_id: this.inferenceId, input, task_type: 'completion' },
+      false,
+      signal
+    );
+    // const usageMetadata = response?.data?.usageMetadata;
+    return response as ChatCompleteResponse;
```

Comment:
The typing on these is not great - I'm an anti-fan of `as` - but looking at the code you're calling, unless ` this.esClient?.inference.inference()` can be set up to return the actual types, I don't think we'd be able to get much better than this, unless we validated the responses with schema ourselves.

Reply:
Good call, I changed that to be more aligned with `InferenceInferenceResult`. There are a dedicated for each task type: `InferenceCompletionResult[]`, `InferenceRankedDocument[]`, `InferenceSparseEmbeddingResult[]`, etc